src/main/java/org/jabref/gui/preferences/NameFormatterTabViewModel.java

```diff
@@ -0,0 +1,97 @@
+package org.jabref.gui.preferences;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javafx.beans.property.ListProperty;
+import javafx.beans.property.SimpleListProperty;
+import javafx.beans.property.SimpleStringProperty;
+import javafx.beans.property.StringProperty;
+import javafx.collections.FXCollections;
+
+import org.jabref.gui.DialogService;
+import org.jabref.model.strings.StringUtil;
+import org.jabref.preferences.JabRefPreferences;
+
+public class NameFormatterTabViewModel implements PreferenceTabViewModel {
+
+    private final ListProperty<NameFormatterItemModel> formatterListProperty = new SimpleListProperty<>(FXCollections.observableArrayList());
+    private final StringProperty addFormatterNameProperty = new SimpleStringProperty();
+    private final StringProperty addFormatterStringProperty = new SimpleStringProperty();
+
+    private final DialogService dialogService;
+    private final JabRefPreferences preferences;
+
+    NameFormatterTabViewModel(DialogService dialogService, JabRefPreferences preferences) {
+        this.dialogService = dialogService;
+        this.preferences = preferences;
+    }
+
+    @Override
+    public void setValues() {
+        formatterListProperty.clear();
+        List<String> names = preferences.getStringList(JabRefPreferences.NAME_FORMATER_KEY);
+        List<String> formats = preferences.getStringList(JabRefPreferences.NAME_FORMATTER_VALUE);
+
+        for (int i = 0; i < names.size(); i++) {
+            if (i < formats.size()) {
+                formatterListProperty.add(new NameFormatterItemModel(names.get(i), formats.get(i)));
+            } else {
+                formatterListProperty.add(new NameFormatterItemModel(names.get(i)));
+            }
+        }
+    }
+
+    @Override
+    public void storeSettings() {
+        // First we remove all rows with empty names.
+        int i = 0;
+        while (i < formatterListProperty.getValue().size()) {
```

Comment:
You should use an Iterator to remove entries in the for loop (your code might also work but I'm not sure how indicies are reshuffeled if you remove an entry.

https://stackoverflow.com/questions/1196586/calling-remove-in-foreach-loop-in-java

Reply:
IntelliJ had another idea (😉): removeIf